### ISSUE TITLE
NVDA no longer gains focus properly when opening VS Code

### ISSUE THREAD

**Issue by Neurrone** [microsoft/vscode#152291]
Issue Type: <b>Bug</b>

/CC @isidorn

NVDA is completely unable to focus on the control with keyboard focus in VS code when alt tabbing into it from another application. This happens 100% of the time for me.

This is very similar to #28316. When this occurs, Vs Code is completely unusable with NVDA (pressing tab and arrow keys doesn't read anything.

I am completely unable to use Vs Code. Only started happening today when I got a new insiders build. Since I haven't updated NVDA for a couple of months, a recent change in Code is a likely suspect.

VS Code version: Code - Insiders 1.69.0-insider (0b74e15affc1b44a5b7cd671dd0a02d3ed6d51a7, 2022-06-15T05:16:09.187Z)
OS version: Windows_NT x64 10.0.19044
Restricted Mode: No
Remote OS version: Linux x64 5.10.102.1-microsoft-standard-WSL2

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|AMD Ryzen 7 PRO 4750U with Radeon Graphics (16 x 1697)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|31.23GB (15.97GB free)|
|Process Argv|--folder-uri=vscode-remote://wsl+Ubuntu/home/dickson/source/backend-api --remote=wsl+Ubuntu --crash-reporter-id 1304d64e-abdb-456c-beb6-72c33120f731|
|Screen Reader|yes|
|VM|0%|

|Item|Value|
|---|---|
|Remote|WSL: Ubuntu|
|OS|Linux x64 5.10.102.1-microsoft-standard-WSL2|
|CPUs|AMD Ryzen 7 PRO 4750U with Radeon Graphics (16 x 1696)|
|Memory (System)|24.44GB (18.99GB free)|
|VM|0%|
</details><details><summary>Extensions (26)</summary>

Extension|Author (truncated)|Version
---|---|---
jupyter-keymap|ms-|1.0.0
remote-containers|ms-|0.240.0
remote-ssh|ms-|0.82.1
remote-ssh-edit|ms-|0.80.0
remote-wsl|ms-|0.66.3
vscode-remote-extensionpack|ms-|0.21.0
vscode-eslint|dba|2.2.2
dendron|den|0.99.0
dendron-markdown-links|den|0.6.20
dendron-markdown-preview-enhanced|den|0.10.57
dendron-markdown-shortcuts|den|0.12.1
dendron-paste-image|den|1.1.0
dendron-snippet-maker|den|0.1.6
gitlens|eam|12.1.0
prettier-vscode|esb|9.5.0
copilot|Git|1.29.6152
rest-client|hum|0.24.6
rainbow-csv|mec|2.4.0
python|ms-|2022.8.0
vscode-pylance|ms-|2022.6.20
jupyter|ms-|2022.5.1001601848
vsliveshare|ms-|1.0.5625
vsliveshare-audio|ms-|0.1.91
vsliveshare-pack|ms-|0.4.0
vscode-mermaid-editor|tom|0.14.0
vscode-remark|uni|1.3.0

</details>

**Comment by Neurrone** [comment]
Switching to stable (v1.66.2) allows me to continue working, so this must have been a very recent regression.

**Comment by isidorn** [comment]
@Neurrone thanks for reporting this.

This sounds like it is connected to the Electron 17->18 update. 
@deepak1556 can you please check?
I know @Neurrone can tryout some builds to help trim down the issue.

**Comment by Neurrone** [comment]
Any chance the Electron upgrade could be reverted before the July 2022 release? It would be a massive issue for screen reader users if it makes it to stable.

**Comment by Neurrone** [comment]
@deepak1556 thanks, that fixed it. Sounds like an upstream bug with Chrome or Electron?

**Comment by isidorn** [comment]
I just tried to reproduce this and could not on my Windows VM with NVDA.
@Neurrone do you have more precise repro steps? Here is what I am doing

1. Start NVDA
2. Start VS Code insiders
3. Alt + Tab into VS Code -> NVDA nicely reads VS Code editor content

Repeat step 3 multiple times -> all is good.
Are you doing something differently?

**Comment by Neurrone** [comment]
@isidorn that's essentially what I did. I've attached a video that demonstrates the problem. Notice that after I alt tab from Vs Code to Firefox, then back into code, I get no feedback at all even though I am moving my cursor.

https://user-images.githubusercont@isidorn nt.com/16471962/176177670-ce1395d5-4f6b-477b-9358-781dbf64cacc.mp4

**Comment by Neurrone** [comment]
Not sure why the link doesn't work, let me try re-uploading again.


https://user-images.githubusercontent.com/16471962/176178689-da67a96a-b230-4db3-96d4-bdbdb5841df6.mp4

**Comment by jeanbrazcosta** [comment]
Hello everybody.

@isidorn I'm also having the same problem as @Neurrone if I go to any other window and go back to vs code insiders I can't read anything else, to get it working again I have to press alt + space.

**Comment by Neurrone** [comment]
This affects Jaws as well. I've seen similar reports of this on other mailing lists.

**Comment by brian-upshaw** [comment]
I'm another user using NVDA with VS Code Insiders (Windows 10) and am being affected by this issue as well.  It tends to not be instantaneous for me; i have to shift focus to another application for more than a few seconds in order for it to trigger, but it is almost any time.  Typically noticed when i alt-tab to a web browser to check documentation.  By the time I switch focus back to VS Code Insiders this has occurred.

**Comment by Neurrone** [comment]
Unfortunately, this appears to have made it into stable. See #154398.

**Comment by jareds** [comment]
I've reproduced this twice in a row with the following steps.
1. Launch VS Code in the current directory using wsl. In my case this is a go project.
2. Open up a text file.
3. Switch to a website in Microsoft Edge.
4. Wait 30 minutes then alt tab back to VS Code.
5. Nothing reads anymore.
Over a period of about 15 minutes, I switched back and forth from VS Code to Edge spending three minutes or less in Edge each time. The only way I could reproduce the issue was by waiting 30 minutes to switch back.
I'm running NVDA 2022.1.
VS Code version info is
Version: 1.69.0 (system setup)
Commit: 92d25e35d9bf1a6b16f7d0758f25d48ace11e5b9
Date: 2022-07-07T05:28:36.503Z
Electron: 18.3.5
Chromium: 100.0.4896.160
Node.js: 16.13.2
V8: 10.0.139.17-electron.0
OS: Windows_NT x64 10.0.19044

**Comment by TheQuinbox** [comment]
Yup, I have this issue too. This has made the editor almost unusable for me.

**Comment by steverep** [comment]
I have the same issue, and adding the `--disable-features=CalculateNativeWinOcclusion` flag does appear to be a viable workaround.  Note though that starting VS Code from WSL will not pass this flag to Electron.  You have to start directly from Windows via PowerShell or command prompt.

**Comment by zersiax** [comment]
Confirmed broken in latest stable. Reverting to v1.68.1 or using the  recommended flag does appear to fix, but this is obviously not good for UX. Do we have an upstream bug for Electron/Chromium given this is likely where this originated? This will likely hit other Electron-using apps, notably Discord, Spotify, Skype, Slack etc.

**Comment by pranavlal** [comment]
Hi all,
I am having the same issue. Here are the steps to duplicate it.
1. With NVDA running, launch visual studio code.
2. Open a file or do anything else. I am writing code on a Linux box from windows so am using the remote extension to work on the box over ssh.
3. Swap out to any other application like edge, outlook etc.
4. Swap back to visual studio code.
The expected result is that I should be able to continue editing.
What happens is that when I use the up or down arrow keys to navigate the text of the editor, I get no speech feedback.. The same happens when using the left or right arrow keys. I am unable to use the menus either. 

The current fix is to use the --disable-features=CalculateNativeWinOcclusion flag.

System Specifications
nvda version= 2022.1
Version: 1.69.0 (user setup)
Commit: 92d25e35d9bf1a6b16f7d0758f25d48ace11e5b9
Date: 2022-07-07T05:28:36.503Z
Electron: 18.3.5
Chromium: 100.0.4896.160
Node.js: 16.13.2
V8: 10.0.139.17-electron.0
OS: Windows_NT x64 10.0.22000

**Comment by steverep** [comment]
The `--disable-features=CalculateNativeWinOcclusion` workaround does not seem to survive locking Windows or switching users.  If I lock my screen or let it lock automatically and come back, I'm back to zero speech in VS code.

**Comment by ashleygrobler04** [comment]
Yeah, got the same issue. 1.68 works perfectly, 1.69 appears to be an issue. Would it work if I switched to insiders somehow?

**Comment by bramd** [comment]
Same issue here. I couldn't find an upstream bug in Electron for this yet. I think it is crucial to get it fixed there before other major apps update their Electron dependency. Unfortunately, I don't know enough technical details to file that bug myself.

**Comment by bkoray** [comment]
Another +1 here, seems like bug made its way to stable. Currently both stable and latest Insiders as of this comment suffers from the issue. I've tried everything including removing all extensions yet issue still persists.

Generally I need to be in another window for some time for it to occur and only way to recover seems to be restarting the editor.

**Comment by mazen428** [comment]
Another +1 here as well.

**Comment by deepak1556** [comment]
**Update:** The issue has been identified and a fix is being worked on currently. I will update here once builds are available for testing. Thanks for your patience!

**Comment by justas2481** [comment]
I have the same problem after update. It is really crytical issue for the blind users, because it makes VSCode impossible to use and we actually rely on VSCode very much, it is the best editor for us. Temporary measure to keep it working is to downgrade or to use command in your CMD "code --disable-features=CalculateNativeWinOcclusion". I hope this will be fixed soon, my job depends on it. P.s. thank you Microsoft team for being responsible and making stuff accessible.
@deepak1556

**Comment by leonardder** [comment]
Wouldn't it be possible to disable `CalculateNativeWinOcclusion` when screen reader mode is on and add that to a patch release?

**Comment by deepak1556** [comment]
*Steps for verification:*

* Install NVDA
* Open insiders
* Ensure editor is screen reader optimized by toggling `Show: Accessibility Help (Alt + F1)` and `Ctrl + E`
* Verify NVDA is able to read editor contents
* Press `Win key + D` to go back to desktop
* From desktop `Alt + Tab` into Insiders
* Verify NVDA is able to read editor contents

**Comment by snufas** [comment]
@isidorn
we had same exact issue a while back,, seems it popped up again.

@deepak1556
I assume the fix will be available in 1 of the nightly builds right?
I only use stable releases, but this issue is such important is that I am waiting for nightly build to be implemented with this fix.

The workout for me currently is just type  any symbol, it could be a space, and then delete that, and the focus will go back to normal. Of course, it's a bit annoying, but for the time being it works.
Please ship that fix ASAP, because some people are relying on this tool alone to get the job done.

**Comment by deepak1556** [comment]
Yes the fix will be shipped first in nightly insiders. Once it has been verified, we already have plan for a recovery release `1.69.1` this week and the fix will be included as part of it.

**Comment by VSCodeTriageBot** [comment]
This bug has been fixed in the latest release of [VS Code Insiders](https://code.visualstudio.com/insiders/)!

@Neurrone, you can help us out by commenting `/verified` if things are now working as expected.

If things still don't seem right, please ensure you're on version 9925e5d4d3d355758cd09d26868f0c12a0006828 of Insiders (today's or later - you can use `Help: About` in the command palette to check), and leave a comment letting us know what isn't working as expected.

Happy Coding!

**Comment by leonardder** [comment]
@deepak1556 Slightly offtopic, but:

* Is it likely that other Microsoft software that's Electron based (e.g. Microsoft Teams) will suffer from this bug as well as benefit from the changes made in the VS specific build of Electron?
* How about Electron itself? Will changes be provided as a pr to upstream as well, or is this not applicable?

**Comment by deepak1556** [comment]
> Is it likely that other Microsoft software that's Electron based (e.g. Microsoft Teams) will suffer from this bug as well as benefit from the changes made in the VS specific build of Electron?

Yup indeed they also hit this issue once updated to Electron >= 18 and the fix will address the issue for other build flavors as well.

> How about Electron itself? Will changes be provided as a pr to upstream as well, or is this not applicable?

Upstream Electron is not affected by the issue. The issue only affects Microsoft builds of Electron due to a certain patch carried internally, I can provide context of the bug if interested but it will dive into chromium internals hence didn't add them here initially.

**Comment by Neurrone** [comment]
@deepak1556 I've verified that the fix works for me as expected. Thanks a lot for prioritizing this fix 🙏 

I'm also relieved that this isn't an upstream issue, as this would be disastrous given how widespread Electron's use is.

**Comment by deepak1556** [comment]
Thanks for confirming!

**Comment by justas2481** [comment]
@deepak1556 I can as well verify that this bug is fixed on Insiders.

**Comment by leonardder** [comment]
I can also confirm that it works like a charm again.

**Comment by marlon-sousa** [comment]
Hello,

Please provide details of this bug if possible, even if it somehow touches chromium internals.

It would be also nice to know when the release recovery will ship (nsome people might not be authorized to use preview).

Thanks,
Marlon

**Comment by jeanbrazcosta** [comment]
I also checked and for me it was fixed.

Em seg., 11 de jul. de 2022 às 10:35, VSCode Triage Bot <
***@***.***> escreveu:
